Features/decrosstalk/ticket 127 clean

danielsf opened the pull request:
This is the initial implementation of the decrosstalking pipeline based on this prototype

https://github.com/AllenInstitute/mesoscope_ica

The design document can be found here

http://confluence.corp.alleninstitute.org/display/IT/Crosstalk+correction+design+document

Currently, the unit test for the entire pipeline (`test_pipeline.py`) just runs the pipeline on a small subset of real data and verifies that the expected outputs are written. There are still some failure modes in the code (which are also failure modes in the prototype) which we have not decided how to handle. After these are fully explored, we can try to write more comprehensive tests, if that seems worthwhile.

**Note:** the trace extraction code in `roi_masks.py` was just copied over from AllenSDK so that this module would not have to import AllenSDK. I did not put a lot of thought into where in this repo the code should live. I am open to feedback on that question.

aamster commented on tests/decrosstalk/data/ophys_plane_instantiation_data.json:
```diff
@@ -0,0 +1,2766 @@
+{
```
I think the naming convention for test data is to place it in a folder called `resources` instead of `data`

djkapner commented on src/ophys_etl/decrosstalk/decrosstalk_schema.py:
```diff
@@ -0,0 +1,121 @@
+import argschema
+
+
+class RoiSchema(argschema.ArgSchema):
```
I think you're using this just as nested, rather than through a parser. It will work, but, `argschema.schemas.DefaultSchema`  is more appropriate.
Same for the next 2 schemas as well.
https://github.com/AllenInstitute/argschema/blob/afc2b9241176a587e543f71830f2968d7a552cb7/argschema/schemas.py#L5

danielsf replied:
I've updated all of the "internal" schemas to inherit from `argschema.DefaultSchema`

aamster commented on tests/decrosstalk/tmp/.gitignore:
```diff
@@ -0,0 +1 @@
+# empty file
```
Did you meant to include `/tmp` in the repo?

danielsf replied:
I did originally, but it has since been removed.

djkapner commented on src/ophys_etl/decrosstalk/decrosstalk_schema.py:
```diff
@@ -0,0 +1,121 @@
+    # the lists do, indeed, need to contain ints.
+
+    decrosstalk_raw_exclusion_label = argschema.fields.List(
+                                       argschema.fields.Field,
```
another way could be to use a marshmallow pre_load decorated function to check inputs before parsing.

danielsf replied:
I'm still very unfamiliar with how marshmallow and argschema work. Given that this function meets the need, do you mind if I leave it as-is? We can revisit this in a future ticket if necessary.

djkapner replied:
it works, leave it

djkapner commented on src/ophys_etl/decrosstalk/decrosstalk_utils.py:
```diff
@@ -0,0 +1,167 @@
+
+    Returns
+    -------
+    A dict
```
I'd recommend using a `TypedDict` defintion for passed or returned dictionaries. It is a little more explicit, can be used in type hinting (which is part of the informatics style guidelines) and can save space if you keep using the same structure.

djkapner commented on src/ophys_etl/decrosstalk/decrosstalk_utils.py:
```diff
@@ -0,0 +1,167 @@
+
+    Parameters
+    -----------
+    trace_dict contains the traces to be validated. It has a structure like:
```
same comment as above about `TypedDict`

aamster commented on src/ophys_etl/decrosstalk/active_traces.py:
```diff
@@ -0,0 +1,465 @@
+import numpy as np
```
Can you remind me where this module came from? I'm wondering if it should be imported from a separate repo rather than included in this repo, as it doesn't really conform to the rest of the code and might be hard to maintain.

danielsf replied:
The file was cut-and-pasted from here

https://github.com/AllenInstitute/mesoscope_ica/blob/main/mesoscope_ica/active_traces.py

(I then reformatted it to get it to pass the linter, but I have not changed any logic)

Based on a conversation I had with Natalia, it's a combination of code written by Allen researchers and outside researchers. It's pretty fine tuned (to the point where, even though the 3 in the denominator on line 134 cannot be justified from first principles, we did not want to change it because the code is tuned to give the correct results for real data). My long term goal is to replace it with code from the event detection pipeline Dan is writing, once that is mature enough to use here.

aamster commented on src/ophys_etl/decrosstalk/ica_utils.py:
```diff
@@ -0,0 +1,197 @@
+        if alpha > 0.3 or beta > 0.3 or alpha < 0 or beta < 0:
+            # Unmixing
+            ica = FastICA(whiten=False, max_iter=10000, random_state=rng)
+            ica.fit_transform(Ow)  # Reconstruct sources
```
This line reads weird to me because it is not an inplace operation. I would expect it to return the result of `FastICA` like 
```
X = FastICA(...)
```
I believe it could be written this way if we addressed the issue on the sklearn side regarding the mixing matrix.

danielsf replied:
Done

danielsf commented on src/ophys_etl/decrosstalk/ica_utils.py:
```diff
@@ -0,0 +1,197 @@
+    while not roi_demixed and it <= iters:
+        if alpha > 0.3 or beta > 0.3 or alpha < 0 or beta < 0:
+            # Unmixing
+            ica = FastICA(whiten=False, max_iter=10000, random_state=rng)
```
sklearn already knows about the bug

https://github.com/scikit-learn/scikit-learn/issues/17162

(though, there doesn't seem to be an attached PR)

My inclination is to leave our fix here because it works. When sklearn addresses that issue, we can compare results to see if they fixed it the way we need them to. I'm not sure I'd be comfortable running our production system off of a fork of sklearn while we wait for them to accept a PR.

djkapner replied:
can we add a `# NOTE` to that effect in the code?

djkapner commented on tests/decrosstalk/test_io_utils.py:
```diff
@@ -0,0 +1,46 @@
+    data['second_dir']['sub3']['b'] = np.array([1.2, 4.3])
+    data['second_dir']['sub3']['c'] = True
+
+    tmp_dir = get_tmp_dir()
```
i don't think this should be used. You are assuming you have write access to the `tmp` dir in the repo. pytest has the `tmp_path` fixture for such things:
https://docs.pytest.org/en/stable/tmpdir.html

djkapner replied:
also, should eliminate need for `teardown_function`

danielsf replied:
How bad is the assumption that whoever is running the tests can write into `tests/decrosstalk/tmp/`?

I'm playing around with
```
import pytest

def test_junk(tmpdir):
    tmpdir.mkdir()
    pass
```

It is producing a weird "File exists" error, both on my Macbook and on synapse. pytest claims that the bug is fixed

https://github.com/pytest-dev/pytest/issues/3451

Obviously not, though.

djkapner replied:
bad enough that I have had this come up in the past.
`tmp_path` rather than `tmpdir` I think will take care of all of this for you.
Example:
https://github.com/AllenInstitute/ophys_etl_pipelines/blob/94d8610013dae5f1a4e67249b5654bd64c13308b/tests/transforms/test_suite2p_wrapper.py#L28-L42

djkapner replied:
I can't recall the exact failure mode, but, I think it had something to do with certain priveleges while building a conda env inside of a dockerfile, and then wanting to run the tests inside the built docker image.

danielsf replied:
Thanks for the help. I obviously don't have a lot of experience with pytest.

I've implemented this change.

djkapner commented on src/ophys_etl/decrosstalk/active_traces.py:
```diff
@@ -0,0 +1,465 @@
+           'find_event_gaps']
+
+
+def mode_robust(input_data, axis=None, d_type=None):
```
have you performed any check that `scipy.stats.mode` doesn't fit the bill?

djkapner replied:
no type hint on return val, incomplete docstring.

djkapner commented on src/ophys_etl/decrosstalk/active_traces.py:
```diff
@@ -0,0 +1,465 @@
+    return data_mode
+
+
+def evaluate_components(traces, event_kernel=5, robust_std=False):
```
no type hint on return val, docstring different format from style guidelines

djkapner commented on src/ophys_etl/decrosstalk/active_traces.py:
```diff
@@ -0,0 +1,465 @@
+    return idx_components, fitness, erfc
+
+
+def find_event_gaps(evs):
```
no type hinting, mixing docstring formats

djkapner replied:
same for later functions

danielsf replied:
I've updated the docstrings of the methods I've touched.

The others were copied from what I'm told is finely tuned code from scientific experts. I'm loath to touch those, since I haven't really altered the behavior of that code and want to preserve the original authors' intent as much as possible.

djkapner replied:
ok

djkapner commented on src/ophys_etl/decrosstalk/decrosstalk_schema.py:
```diff
@@ -0,0 +1,121 @@
+                      description='order of group in session',
+                      required=True)
+
+    planes = argschema.fields.List(
```
you might also consider Nested(many=True)
https://marshmallow.readthedocs.io/en/stable/marshmallow.fields.html#marshmallow.fields.Nested
example:
https://github.com/AllenInstitute/ophys_etl_pipelines/blob/1225bd042c8e51598dbbfd30ca26ab6b27f276e1/src/ophys_etl/transforms/classification.py#L233-L237

djkapner commented on src/ophys_etl/decrosstalk/io_utils.py:
```diff
@@ -0,0 +1,427 @@
+        out_file.write(json.dumps(data, indent=2, sort_keys=True))
+
+
+# The classes below are implemented to allow us to quickly
```
There is now a LIMS strategy for this. Can we just include ouptut formatting for what is expected in production? It seems like if we had to, we/the science team could write translators to the old format.

danielsf replied:
This block of code came about when it became apparent that it would take less work for me to change the output of the module to fit what Natalia's plotting code needed than it would take for Natalia to change her plotting code. Once we have ownership of the plotting code we can think more clearly about what these files should look like. Given that none of this code affects the outputs that LIMS expects (these classes all exist to populate the QC outputs directory), I would like to just leave this as-is and revisit the question of what should be done here in a future ticket.

Is that reasonable?

djkapner replied:
Reasonable

djkapner commented on src/ophys_etl/decrosstalk/ophys_plane.py:
```diff
@@ -0,0 +1,848 @@
+        Load the data from self._path; store te data in self._data
+        """
+        with h5py.File(self.path, mode='r') as in_file:
+            self._data = in_file['data'][()]
```
It is the h5py way to read all the data in a dataset:
https://docs.h5py.org/en/stable/high/dataset.html#reading-writing-data

djkapner commented on src/ophys_etl/decrosstalk/ophys_plane.py:
```diff
@@ -0,0 +1,848 @@
+
+        return final_output
+
+    def get_raw_traces(self, other_plane):
```
can be outside the class or a classmethod. easier to test, less `classy`

djkapner commented on src/ophys_etl/decrosstalk/ophys_plane.py:
```diff
@@ -0,0 +1,848 @@
+
+        return output
+
+    def unmix_ROI(self, roi_traces, seed=None, iters=10):
```
outside the class or a classmethod

djkapner commented on src/ophys_etl/decrosstalk/ophys_plane.py:
```diff
@@ -0,0 +1,848 @@
+        return output
+
+
+class OphysPlane(object):
```
This class gives me pause. First, it's name is very general, making me think it is a general data structure for a plane. But, then it has methods specific to decrosstalk. I'd be more inclined to define strict data-only structures and stand-alone methods that accept and return those data structures.
We get a lot of classes like this from the science team and they can be tough to test and maintain because somewhere deep in a method there is a `self.some_important_parameter` that is set by `self.some_other_method()` which isn't very transparent.
I understand it is a big ask to refactor this with time constraints, but, something to keep in mind later.

danielsf replied:
Really, the `OphysPlane` class just exists to keep and `OphysMovie` and a list of `OphysROIs` together. You are correct that all of the processing methods could be pulled out of the class, as long as `run_decrosstalk` accepts two `OphysPlane` instantiations.

How would you feel about renaming the class `DecrosstalkingPlane` and then refactoring the methods as described (`run_decrosstalk` accepts two `DecrosstalkingPlane`s; everything else remains the same, but outside of any class)?

danielsf replied:
@njmei
I'm pinging you because this initial design was meant to satisfy your request for something that looked like
```
plane1 = OphysPlane(data1)
plane2 = OphysPlane(data2)

plane1.run_decrosstalk(plane2)
plane2.run_decrosstalk(plane1)
```

I assume something that looks like

```
plane1 = OphysPlane(data1)
plane2 = OphysPlane(data2)

run_decrosstalk(plane1, plane2)
run_decrosstalk(plane2, plane1)
```
won't offend your sensibilities?

njmei replied:
@danielsf yes, that works for me.

djkapner replied:
I'd feel better about it, but this isn't a showstopper for me. The class is not too gnarly, it is just a style that eventually gets gnarly, if persisted.

danielsf replied:
About to push the fix

djkapner commented on src/ophys_etl/transforms/decrosstalk_wrapper.py:
```diff
@@ -62,7 +62,9 @@ def run(self):
                 for k in ('roi', 'neuropil'):
+                    out_fname = p0['output_%s_trace_file' % k]
                     if k not in traces_0:
+                        print('skipping %s' % out_fname)
```
`ArgSchemaParser` class include a logger, preferable to prints. Example:
https://github.com/AllenInstitute/ophys_etl_pipelines/blob/94d8610013dae5f1a4e67249b5654bd64c13308b/src/ophys_etl/transforms/postprocess_rois.py#L99-L103

danielsf replied:
Don't worry. These are all getting `revert`ed once I finish debugging my latest chicanery.

djkapner replied:
gotcha

danielsf commented on src/ophys_etl/decrosstalk/decrosstalk_schema.py:
```diff
@@ -0,0 +1,121 @@
+                             description='experiment ID',
+                             required=True)
+
+    motion_corrected_stack = argschema.fields.Str(
```
I've updated the field to be an `InputFile`.

Because the name of the field involves changes to the LIMS strategy that generates the input.json, I'm going to leave it as `motion_corrected_stack` (I think this is the formal name of that field in the LIMS database)

djkapner replied:
Kat wrote a special `H5InputFile` field. I think it was because `InputFile` was not being properly validated on windows (at that time, Isaak was working in windows.) It doesn't affect production, but, it is here:
https://github.com/AllenInstitute/ophys_etl_pipelines/blob/94d8610013dae5f1a4e67249b5654bd64c13308b/src/ophys_etl/schemas/fields.py#L5-L20

danielsf commented on src/ophys_etl/decrosstalk/decrosstalk_schema.py:
```diff
@@ -0,0 +1,121 @@
+                        description='path to motion corrected movie',
+                        required=True)
+
+    motion_border = argschema.fields.Dict(description='motion border',
```
Naively trying to populate those lists raised a bunch of marshmallow errors. Given that we still need to evaluate all of the edge cases in the real data, do you mind if I leave this for a future ticket? `motion_border` is a field in LIMS with a known structure, so the meaning of this field is not totally nebulous.

danielsf commented on src/ophys_etl/decrosstalk/decrosstalk_schema.py:
```diff
@@ -0,0 +1,121 @@
+                           description='ophys_session_id',
+                           required=True)
+
+    qc_output_dir = argschema.fields.Str(
```
Done

danielsf commented on src/ophys_etl/decrosstalk/decrosstalk_utils.py:
```diff
@@ -0,0 +1,167 @@
+def validate_traces(trace_dict):
+    """
+    Check a traces_dict for validity.
+    Validity is defined as neuropile and roi traces having the same shape.
```
fixed

danielsf commented on src/ophys_etl/decrosstalk/ophys_plane.py:
```diff
@@ -0,0 +1,848 @@
+        """
+
+        ica_input = np.array([roi_traces['signal'], roi_traces['crosstalk']])
+        assert ica_input.shape == (2, len(roi_traces['signal']))
```
This was actually something I included to catch bugs as I was developing. I'm just going to remove it.

danielsf commented on src/ophys_etl/decrosstalk/roi_masks.py:
```diff
@@ -0,0 +1,551 @@
+
+class Mask(object):
+    '''
+    Abstract class to represent image segmentation mask. Its two
```
I literally cut-and-paste this code from AllenSDK so that we could do trace extraction without importing from AllenSDK. Do you mind if I declare sanitizing this file "out of scope" for this ticket? At some time in the future, we are going to want a more reasoned approach to doing trace extraction in this repo, anyway.

danielsf commented:
@djkapner @aamster 

I think I have addressed your concerns with the notable exceptions of:

 - Typing: I've never worked with type-hinted python before and am going to need to do some reading about this
 - Validating the schema: also haven't worked much with marshmallow and argschema, so the learning curve here will slow me down
 
 I'll try to learn about type hinting tonight so that can be the final push. I'd like to punt on the schema validation for now, since what we have works.
 
Do you mind taking a look at this tomorrow morning to see if there are any other glaring showstoppers?

At some point I'm going to need to move on to running problem experiments through the code to see if they are still problems.
 
 Thanks

djkapner commented:
> @djkapner @aamster
> 
> I think I have addressed your concerns with the notable exceptions of:
> 
>     * Typing: I've never worked with type-hinted python before and am going to need to do some reading about this
> 
>     * Validating the schema: also haven't worked much with marshmallow and argschema, so the learning curve here will slow me down
> 
> 
> I'll try to learn about type hinting tonight so that can be the final push. I'd like to punt on the schema validation for now, since what we have works.
> 
> Do you mind taking a look at this tomorrow morning to see if there are any other glaring showstoppers?
> 
> At some point I'm going to need to move on to running problem experiments through the code to see if they are still problems.
> 
> Thanks

will take a look tomorrow. Here's a nice example where Taylor defined a TypedDict:
https://github.com/AllenInstitute/ophys_etl_pipelines/blob/795bff36ed6d4b7e5788c5d7e7ce71c78d126bbf/src/ophys_etl/transforms/sine_dewarp.py#L18-L36
which then made referring to it in docstrings and typehints nice and smooth:
See the functions that follow it, including this one, where it is the returned object:
https://github.com/AllenInstitute/ophys_etl_pipelines/blob/795bff36ed6d4b7e5788c5d7e7ce71c78d126bbf/src/ophys_etl/transforms/sine_dewarp.py#L255-L301

danielsf commented:
@djkapner 
The dicts I'm trying to type look like
```
MyDict['roi'][roi_id]['signal']
MyDict['roi'][roi_id]['crosstalk']
```
Where `roi_id` can be an arbitrary `int`. I cannot find a way to write a `TypedDict` that accepts an unspecified key. I can imagine a scheme where we define a bunch of wrappers around `dict`s that validate keys and values have the right type when we invoke `__setitem__`.  Something like:

```
class NeuronDict(object):

    def __init__(self):
        self._data = {}

    def __setitem__(self, key:int, value:NeuronClass):
        if not isinstance(key, int):
            raise error
        if not isinstance(value, NeuronClass):
            raise error
        self._data[key] = value
```

That is several hours of work. There are several edge cases in the data that I need to investigate during tomorrow's QCathon.

Do you mind if we put a pin in this discussion for the moment and come back to the typing of the methods in this module in a later ticket, which I can hopefully get to on Friday or early next week, after I've investigated some of those datasets?

djkapner commented:
> @djkapner
> The dicts I'm trying to type look like
> 
> ```
> MyDict['roi'][roi_id]['signal']
> MyDict['roi'][roi_id]['crosstalk']
> ```
> 
> Where `roi_id` can be an arbitrary `int`. I cannot find a way to write a `TypedDict` that accepts an unspecified key. I can imagine a scheme where we define a bunch of wrappers around `dict`s that validate keys and values have the right type when we invoke `__setitem__`. Something like:
> 
> ```
> class NeuronClass(object):
> 
>     self._data['signal'] = some array
>     self._data['crosstalk'] = some array
> 
>     ... methods to access 'signal' and 'crosstalk'...
> 
> class NeuronDict(object):
> 
>     def __init__(self):
>         self._data = {}
> 
>     def __setitem__(self, key:int, value:NeuronClass):
>         if not isinstance(key, int):
>             raise error
>         if not isinstance(value, NeuronClass):
>             raise error
>         self._data[key] = value
> ```
> 
> That is several hours of work. There are several edge cases in the data that I need to investigate during tomorrow's QCathon.
> 
> Do you mind if we put a pin in this discussion for the moment and come back to the typing of the methods in this module in a later ticket, which I can hopefully get to on Friday or early next week, after I've investigated some of those datasets?

yep, makes sense to pick this up later

djkapner reviewed (approved):
Approved, with the hope that:
* some commit squashing occurs before merging
* new issue(s) written to list loose ends we don't have time for right now. (mostly aesthetics and style)
